crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl

```diff
@@ -0,0 +1,218 @@
+
+#import bevy_core_pipeline::fullscreen_vertex_shader::FullscreenVertexOutput
+#import bevy_pbr::mesh_view_bindings::{lights, view}
+#import bevy_pbr::mesh_view_types::DIRECTIONAL_LIGHT_FLAGS_VOLUMETRIC_BIT
+#import bevy_pbr::shadow_sampling::sample_shadow_map_hardware
+#import bevy_pbr::shadows::{get_cascade_index, world_to_directional_light_local}
+#import bevy_pbr::view_transformations::{
+    frag_coord_to_ndc,
+    position_ndc_to_view,
+    position_ndc_to_world
+}
+
+// The GPU version of [`VolumetricFogSettings`]. See the comments in
+// `volumetric_fog/mod.rs` for descriptions of the fields here.
+struct VolumetricFog {
+    fog_color: vec3<f32>,
+    light_tint: vec3<f32>,
+    ambient_color: vec3<f32>,
+    ambient_intensity: f32,
+    step_count: u32,
+    max_depth: f32,
+    absorption: f32,
+    scattering: f32,
+    density: f32,
+    scattering_asymmetry: f32,
+    light_intensity: f32,
+}
+
+@group(1) @binding(0) var<uniform> volumetric_fog: VolumetricFog;
+@group(1) @binding(1) var color_texture: texture_2d<f32>;
+@group(1) @binding(2) var color_sampler: sampler;
+
+#ifdef MULTISAMPLED
+@group(1) @binding(3) var depth_texture: texture_depth_multisampled_2d;
+#else
+@group(1) @binding(3) var depth_texture: texture_depth_2d;
+#endif
+
+// 1 / (4π)
+const FRAC_4_PI: f32 = 0.07957747154594767;
+
+// The common Henyey-Greenstein asymmetric phase function [1] [2].
+//
+// This determines how much light goes toward the viewer as opposed to away from
+// the viewer. From a visual point of view, it controls how the light shafts
+// appear and disappear as the camera looks at the light source.
+//
+// [1]: https://www.scratchapixel.com/lessons/3d-basic-rendering/volume-rendering-for-developers/ray-marching-get-it-right.html
+//
+// [2]: https://www.pbr-book.org/4ed/Volume_Scattering/Phase_Functions#TheHenyeyndashGreensteinPhaseFunction
+fn henyey_greenstein(neg_LdotV: f32) -> f32 {
+    let g = volumetric_fog.scattering_asymmetry;
+    let denom = 1.0 + g * g - 2.0 * g * neg_LdotV;
+    return FRAC_4_PI * (1.0 - g * g) / (denom * sqrt(denom));
+}
+
+@fragment
+fn fragment(in: FullscreenVertexOutput) -> @location(0) vec4<f32> {
+    // Unpack the `volumetric_fog` settings.
+    let fog_color = volumetric_fog.fog_color;
+    let ambient_color = volumetric_fog.ambient_color;
+    let ambient_intensity = volumetric_fog.ambient_intensity;
+    let step_count = volumetric_fog.step_count;
+    let max_depth = volumetric_fog.max_depth;
+    let absorption = volumetric_fog.absorption;
+    let scattering = volumetric_fog.scattering;
+    let density = volumetric_fog.density;
+    let light_tint = volumetric_fog.light_tint;
+    let light_intensity = volumetric_fog.light_intensity;
+
+    let exposure = view.exposure;
+
+    // Sample the depth. If this is multisample, just use sample 0; this is
+    // approximate but good enough.
+    let frag_coord = in.position;
+    let depth = textureLoad(depth_texture, vec2<i32>(frag_coord.xy), 0);
```

Comment:
This won't work on webgl but that's not really a big issue since this feature is optional anyway. We might want to warn about that inside the plugin initialization. Not a blocker for merging though. Also, the fact that we can't sample depth in webgl is most likely a wgpu issue and not a webgl issue.